### PR TITLE
Indexing and POS abbreviations(subtag) mapping

### PR DESCRIPTION
**Pull request by pnklshh**
Combined both indexing and mapping in a single pull request

1. Storing indexes in file and retrieving on start of the application, use line ids from indexing to lookup for the faster in DB
2. Mapping subtags of Brown corpus to generalized tags for English language

### CODE REVIEW

File: backend.py [hunk 1]
```diff
@@ -2,41 +2,62 @@
...
+app.config['MYSQL_DATABASE_USER'] = 'pnkls'
+app.config['MYSQL_DATABASE_PASSWORD'] = 'root'
 app.config['MYSQL_DATABASE_DB'] = 'wordsense'
 
 mysql = MySQL(app)
 conn = mysql.connect()
 cursor = conn.cursor()
 
+def GetIndexing():
+    f = open("indexing.txt", "r")
```
Comment (oldosamaibraheem11) [inline backend.py]: We should change f to file

Comment (oldosamaibraheem11) [reply backend.py]: We should also think about moving this to the logic file. The reason is because the backend.py file is there to hold the url directs

File: backend.py [hunk 2]
```diff
@@ -2,41 +2,62 @@
...
+    dictionary = json.loads(f.read())
+    return dictionary
+
 @app.route('/', methods =["GET", "POST"])
 def homepage():
+    # Get the indexing in memory so that we have it until the application is closed
+    dictionary = GetIndexing()
```
Comment (oldosamaibraheem11) [inline backend.py]: This is a good idea, so then we don't have to make calls everytime to get the index

File: backend.py [hunk 3]
```diff
@@ -2,41 +2,62 @@
...
         error = ""
     if request.method == "POST":
         language_selected = request.form.get('language')
         Lang_ID = logic.SQLQuery(f"SELECT Lang_ID FROM Lang_Ref WHERE Lang_Desc = '{language_selected}';")
         part_of_speech_selected = request.form.get('partOfSpeech')
-        part_of_speech_ID = logic.SQLQuery(f"SELECT Part_ID FROM Speech_Parts WHERE Lang_ID = '{Lang_ID[0]}' AND Part_Desc = '{part_of_speech_selected}';")
-        word_selected = logic.wordcreator(request.form.get('word'),request.form.get('partOfSpeech'))
-        sentence_List = logic.SQLQuery(f"select Line_Text from {language_selected}_corpus where Line_Text like  '% {word_selected} %';")
+        word_selected = (request.form.get('word') + '/' + request.form.get('partOfSpeech')).lower()
+
```
Comment (oldosamaibraheem11) [inline backend.py]: I am not sure if GitHub is doing or if it is the way the code is written but we should get rid of the blank lines for example line 41

File: backend.py [hunk 4]
```diff
@@ -2,41 +2,62 @@
...
+        line_ids = dictionary[word_selected]
+        # ignore first and last characters i.e. '[' and ']' to get the list of line ids as a string like "1,3,6,7,...""
```
Comment (oldosamaibraheem11) [inline backend.py:21]: We should combine lines 42 and 44 into a function or something like that. We can turn it into one line 
line_ids = str(dictionary[word_selected])[1:][:-1]

File: logic.py [hunk 5]
```diff
@@ -7,15 +7,65 @@
...
+app.config['MYSQL_HOST'] = 'localhost'
+app.config['MYSQL_DATABASE_USER'] = 'pnkls'
+app.config['MYSQL_DATABASE_PASSWORD'] = 'root'
 app.config['MYSQL_DATABASE_DB'] = 'wordsense'
 
 mysql = MySQL(app)
 conn = mysql.connect()
 cursor = conn.cursor()
 
+english_pos_mapping = {
```
Comment (oldosamaibraheem11) [inline logic.py:24]: I like the use of a dictionary structure but what we had this in a csv file. Then we can add and remove it much easier than having to change the code everytime

File: parser.py [hunk 6]
```diff
@@ -9,23 +9,52 @@
 import sys
 from os import listdir
 from os.path import isfile, join
-nltk.download('punkt')
+from collections import defaultdict
+import json
 
 def GetLanguageId(language):
-    language_id_list = logic.SQLQuery("select Lang_ID from lang_ref where Lang_Desc = '"+ language +"'")
+    language_id_list = logic.SQLQuery("select Lang_ID from lang_ref where Lang_Desc = '" + language + "'")
```
Comment (oldosamaibraheem11) [inline parser.py]: We should move this to the logic file and then Parser.py can use it and backend.py. This is a very useful function! good!

File: parser.py [hunk 7]
```diff
@@ -9,23 +9,52 @@
...
 
-    # Condition for language not present in database
     if len(language_id_list) == 0:
         return -1
     else:
         return language_id_list[0]
 
+def StoreIndexing(dictionary):
+    f = open("indexing.txt", "w")
```
Comment (oldosamaibraheem11) [inline parser.py]: again we should change f to a better name

File: parser.py [hunk 8]
```diff
@@ -9,23 +9,52 @@
...
+    # number of parts formed for each tagged word in the corpus(word/pos)
+    tagged_word_parts = 2   
+    
+    for word in word_list:
+        # skip untagged words
+        if(len(word.split('/')) != tagged_word_parts):
+            continue
+
+        subtag = (word.split('/')[1]).upper()
+        # ignore -TL suffix in subtags since they are just to indicate that the word occurs in title
```
Comment (oldosamaibraheem11) [inline parser.py:25]: good naming in this area, could we just fix the formatting of the lines. It can sometimes be hard to follow

**Review by oldosamaibraheem11 — CHANGES_REQUESTED** [review]
See commnets

File: backend.py (hunk 1 above)
Comment (FaizanSyed95) [inline backend.py]: Variable f name should be changed, for good coding. Also, I agree with the suggestion by Osama.

File: backend.py [hunk 9]
```diff
@@ -2,41 +2,62 @@
...
+        line_ids_str = str(line_ids)[1:][:-1]
```
Comment (FaizanSyed95) [inline backend.py]: If there is no use of the variable line_ids other than just to turn into a string, then it's a good idea to do that in one line, just like Osama said.

File: parser.py (hunk 8 above)
Comment (pnklshh) [reply parser.py:25]: Do you mean removing blank lines by formatting of lines ?

File: backend.py [hunk 10]
```diff
@@ -2,41 +2,62 @@
...
+    # Get the indexing in memory so that we have it until the application is closed
+    dictionary = GetIndexing()
+
     if request.method == "GET":
         sentence_List = ""
         error = ""
     if request.method == "POST":
         language_selected = request.form.get('language')
         Lang_ID = logic.SQLQuery(f"SELECT Lang_ID FROM Lang_Ref WHERE Lang_Desc = '{language_selected}';")
         part_of_speech_selected = request.form.get('partOfSpeech')
```
Comment (FaizanSyed95) [inline backend.py]: I think we should add some comments about words like "Lang_ID, Lang_ref, and Land_desc so that a reviewer who doesn't know about the database can understand what these listings are.

File: parser.py (hunk 8 above)
Comment (oldosamaibraheem11) [reply parser.py:25]: Yes

File: backend.py [hunk 11]
```diff
@@ -2,41 +2,62 @@
...
         error = ""
         if len(sentence_List) == 0:
             error = "Error: Word not in corpus"
+
     language_list = logic.SQLQuery("SELECT Lang_Desc FROM Lang_Ref;")
     part_of_speech_list = logic.SQLQuery("SELECT Part_Desc FROM Speech_Parts WHERE Lang_ID = 1;")
     return render_template('index.html', language_list = language_list, part_of_speech_list=part_of_speech_list, sentence_List=sentence_List, error=error)
+
 @app.route('/Query', methods =["GET", "POST"])
 def Query():
```
Comment (FaizanSyed95) [inline backend.py:32]: I thinking if we can add few comments for the working of this function would be a good idea to understand it.

File: logic.py (hunk 5 above)
Comment (FaizanSyed95) [reply logic.py:24]: Yes, I think the same. I think this is what the professor wants as well.

File: parser.py (hunk 7 above)
Comment (FaizanSyed95) [reply parser.py]: I would like to suggest the same thing.

File: parser.py (hunk 8 above)
Comment (FaizanSyed95) [reply parser.py:25]: I think the formatting is okay, it's just the comments that are no looking formatted, it can be an issue of Git.

File: parser.py [hunk 12]
```diff
@@ -9,23 +9,52 @@
...
+    f.write(dictionary)
+    f.close()
+
+def CreateIndexing(sentence, line_id, dictionary):
+    word_list = sentence.split()
+    # number of parts formed for each tagged word in the corpus(word/pos)
+    tagged_word_parts = 2   
+    
+    for word in word_list:
+        # skip untagged words
```
Comment (FaizanSyed95) [inline parser.py:21]: I like this idea, when we will be using untagged corpus we can simply change this.

File: parser.py [hunk 13]
```diff
@@ -9,23 +9,52 @@
...
+        subtag = (word.split('/')[1]).upper()
+        # ignore -TL suffix in subtags since they are just to indicate that the word occurs in title
+        subtag = subtag.split('-TL')[0]
+
+        if(subtag in logic.english_pos_mapping):
+            generalized_pos = logic.english_pos_mapping[subtag]
+        else:
+            generalized_pos = subtag
+        
+        word = (word.split('/')[0] + "/" + generalized_pos).lower()
```
Comment (FaizanSyed95) [inline parser.py:31]: Can you explain, why are we lower casing the words?

**Review by FaizanSyed95 — CHANGES_REQUESTED** [review]
Overall it looks good, just required few changes.

File: backend.py (hunk 3 above)
Comment (pnklshh) [reply backend.py]: I have intentionally added blank lines after some lines so that the code looks cleaner and easier to understand different parts. I can remove them if the team wants.

File: logic.py (hunk 5 above)
Comment (pnklshh) [reply logic.py:24]: Since we decided to add mappings in logic.py file, I added it like a dictionary here. How if we go ahead with this now and if ask others if we want to change.

File: backend.py (hunk 4 above)
Comment (pnklshh) [reply backend.py:21]: Sure

File: backend.py (hunk 1 above)
Comment (pnklshh) [reply backend.py]: Sure, I will do these

File: backend.py (hunk 3 above)
Comment (oldosamaibraheem11) [reply backend.py]: Sounds good

File: logic.py (hunk 5 above)
Comment (oldosamaibraheem11) [reply logic.py:24]: Sounds good

File: parser.py (hunk 7 above)
Comment (pnklshh) [reply parser.py]: Noted.

File: logic.py (hunk 5 above)
Comment (FaizanSyed95) [reply logic.py:24]: sounds good to me too.

File: parser.py (hunk 13 above)
Comment (pnklshh) [reply parser.py:31]: Since a user may enter the word in random case...so we convert to lower at both ends - searching for the user input word as well as while storing it in the index file.

File: backend.py (hunk 10 above)
Comment (pnklshh) [reply backend.py]: If we add comments to describe each variable that would turn out to be a lot more comments. And I think the variable names are self explanatory, aren't they ?

File: backend.py (hunk 9 above)
Comment (pnklshh) [reply backend.py]: Sure.

File: backend.py (hunk 1 above)
Comment (pnklshh) [reply backend.py]: Yes I changed this.

**Comment by pnklshh** [comment]
Made the changes in another commit...Please check.

**Comment by oldosamaibraheem11** [comment]
@pnklshh Do you have any unit tests written for these functions?

File: backend.py (hunk 10 above)
Comment (FaizanSyed95) [reply backend.py]: I agree, however, I'm asking for some limited variables which are not easy to understand.

File: parser.py (hunk 13 above)
Comment (FaizanSyed95) [reply parser.py:31]: Would be the same for other corpora as well?

File: parser.py [hunk 14]
```diff
@@ -9,26 +9,35 @@
...
+    tagged_word_parts = 2   
+    for word in word_list:
+        # skip untagged words
+        if(len(word.split('/')) != tagged_word_parts):
+            continue
+        subtag = (word.split('/')[1]).upper()
+        # ignore -TL suffix in subtags since they are just to indicate that the word occurs in title
+        subtag = subtag.split('-TL')[0]
+        if(subtag in logic.english_pos_mapping):
+            generalized_pos = logic.english_pos_mapping[subtag]
```
Comment (Garrett-Reed) [inline parser.py:28]: Does this return the generalized tag, or the subtag?

**Review by Garrett-Reed — CHANGES_REQUESTED** [review]
In the future we'll want to move the subtag to general tag mapping to an external file, but overall it looks good. There was one spot where the code might be returning a value we don't expect.

File: parser.py (hunk 14 above)
Comment (pnklshh) [reply parser.py:28]: It returns the generalized tag

**Comment by pnklshh** [comment]
@osamaibraheem11 I do not have written unit tests. Do we need them for this sprint or we are taking them up from next sprint ?

**Comment by oldosamaibraheem11** [comment]
@pnklshh it would helpful to have some unit testing, also Dr. Scannell recommend we start testing this sprint. Look at the testing plan I sent in slack, let me know if you have questions or suggestions
